### PR TITLE
[MXNET-623] Fixing an integer overflow bug in large NDArray

### PR DESCRIPTION
**Pull request by apeforest**
## Description ##
User filed a bug when converting NDArray to numpy. The rootcause is due to the large tensor size that exceeds the current limit (32-bit unsigned integer). This fix is to replace the unsigned 32-bit integer by int64_t data type to index tensor size. 

I verified the fix using the following code from its original [issue](https://github.com/apache/incubator-mxnet/issues/11495):
```
import mxnet as mx
from mxnet import gluon, nd
m = gluon.nn.Embedding(14000, 128)
m.initialize()
ind = nd.zeros((700000, 128))
x = m(ind)
x.shape
test=x.asnumpy()
```

I have run python memory_profiler and did not see noticeable memory footprint increase. I am still 
evaluating the performance in more cases.

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [X] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [X] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage:
- Nightly tests are added for complicated/long-running ones: test_large_array.py
- [X ] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [X ] Changed the shape index from uint32_t to int64_t.

## Comments ##
- This PR is to fix a bug filed on Github. It is not verified to support large array in all operators or scenarios in MXNet. A project is needed to address large array support in a large scope.
- This PR has a dependency on mshadow submodule. The related PR is https://github.com/dmlc/mshadow/pull/348

### CODE REVIEW

**Comment by apeforest** [comment]
@eric-haibin-lin Please help to review this.

**Comment by apeforest** [comment]
@lupesko Thanks for your comments. This PR still depends on another [PR (348)](https://github.com/dmlc/mshadow/pull/348) in mshadow submodule. Therefore, this PR cannot pass the CI until [348](https://github.com/dmlc/mshadow/pull/348) is merged. I plan to address all the required checklists in description when the dependency of this PR is cleared. I have updated the title with [WIP]. I will also add unit test when it is complete for CI.

**Comment by eric-haibin-lin** [comment]
For test with large arrays, it's better to put them under nightly test suites.
cc @marcoabreu

File: src/operator/tensor/ordering_op-inl.h [hunk 1]
```diff
@@ -367,7 +367,7 @@ void TopKImpl(RunContext ctx,
   // Additional temp space for gpu full sorts for batch ids.
   temp_size += sizeof(int) * src.Size();
   // Temp space for cpu sorts.
-  temp_size = std::max(temp_size, sizeof(real_t) * src.Size());
+  temp_size = std::max(temp_size, sizeof(real_t) * static_cast<size_t>(src.Size()));
```
Comment (anirudh2290) [inline src/operator/tensor/ordering_op-inl.h]: we should change this to index_t.

**Comment by eric-haibin-lin** [comment]
@marcoabreu it's a different one

**Comment by access2rohit** [comment]
LGTM .... also address @anirudh2290 comment too.

File: tests/nightly/test_large_array.py [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+    m = gluon.nn.Embedding(14000, 128)
+    m.initialize()
+    ind = nd.zeros((700000, 128))
+    x = m(ind)
+    x.shape
+    test=x.asnumpy()
+    assert (x.shape == test.shape)
+
+if __name__ == '__main__':
+    test_ndarray2numpy()
```
Comment (eric-haibin-lin) [inline tests/nightly/test_large_array.py]: I think you need to update https://github.com/apache/incubator-mxnet/blob/master/tests/nightly/Jenkinsfile to enable nightly test @marcoabreu correct me if i'm wrong

File: src/operator/tensor/ordering_op-inl.h (hunk 1 above)
Comment (apeforest) [reply src/operator/tensor/ordering_op-inl.h]: Updated.

File: tests/nightly/test_large_array.py (hunk 2 above)
Comment (marcoabreu) [reply tests/nightly/test_large_array.py]: To be honest, I can't find any evidence of the fact that we run the python tests in tests/nightly at all right now :(

I'd rather put these tests in tests/python/unittest or gpu. The most preferred method on the other hand (considering edge devices) would be to make a C++ test and then mock the memory allocation. 

@larroy @lebeg might want to chime in

Comment (apeforest) [reply tests/nightly/test_large_array.py]: @marcoabreu Do we have some integration test suite? I can add a C++ unit test but still prefer if we can test this from end-to-end.

**Comment by apeforest** [comment]
There are some compilation issue on Windows and GPU after I updated submodule of mshadow. Change the status of the PR to WIP for now.

File: tests/nightly/test_large_array.py (hunk 2 above)
Comment (marcoabreu) [reply tests/nightly/test_large_array.py]: Sorry, I'm currently swamped with high priority tasks and I don't have time to review your task or assist with inqueries.

**Comment by apeforest** [comment]
@sandeep-krishnamurthy Not able to reproduce the CI test error in a standalone environment with MKLDNN and no CUDNN. Investigating the rootcaust.

**Comment by eric-haibin-lin** [comment]
Will probably help this case, too https://discuss.mxnet.io/t/oom-when-trying-to-slice-and-print-nd-array/1832

**Comment by lebeg** [comment]
@apeforest could you rebase?

File: src/c_api/c_api_function.cc [hunk 3]
```diff
@@ -56,7 +56,7 @@ std::vector<nnvm::NodeEntry> Gradient(
 
   std::vector<nnvm::NodeEntry> ret;
   for (index_t i = 0; i < g->num_outputs(); ++i) {
-    ret.emplace_back(nnvm::NodeEntry{g, i, 0});
+    ret.emplace_back(nnvm::NodeEntry{g, static_cast<uint32_t>(i), 0});
```
Comment (lebeg) [inline src/c_api/c_api_function.cc]: For large values of `i` this cast would fail, right? Probably another change in nnvm is needed to use `index_t` for the index: [node.h](https://github.com/dmlc/nnvm/blob/ef0ab9b09dbf1318851be311d3752de6c9bd4881/include/nnvm/node.h#L35)

Comment (apeforest) [reply src/c_api/c_api_function.cc]: The Node data structure in nnvm::Graph uses uint32_t to index graph nodes. In that case, the number of nodes in the graph is not going to be greater than 2^32. I have communicated this to the tvm team and they suggest we keep the uint32_t for node id.

Comment (lebeg) [reply src/c_api/c_api_function.cc]: @apeforest Ok, but then I think it would make sense to indicate an error here somehow, because if it doesn't fit the behaviour is undefined.

**Comment by vandanavk** [comment]
Thank you for your contributions @apeforest .

Is this still WIP or is pending only the failing CI build?

**Comment by apeforest** [comment]
@vandanavk One unit test failed on CuDNN platform. I was working on the fix but was delayed due to other higher priority tasks. I will investigate this issue again and provide the fix to unit test asap.

**Comment by apeforest** [comment]
@sandeep-krishnamurthy @nswamy This PR is ready for review. Could you please change the label to [pr-awaiting-review]? Thanks!

File: src/io/iter_image_recordio_2.cc [hunk 4]
```diff
@@ -328,7 +328,8 @@ inline bool ImageRecordIOParser2<DType>::ParseNext(DataBatch *out) {
         n_to_out = 0;
       }
     } else {
-      int n_to_copy = std::min(n_parsed_, batch_param_.batch_size - current_size);
+      int n_to_copy = std::min(n_parsed_,
+        static_cast<unsigned>(batch_param_.batch_size) - current_size);
```
Comment (lebeg) [inline src/io/iter_image_recordio_2.cc]: Maybe `current_size` should be `size_t` instead?

File: src/operator/convolution_v1-inl.h [hunk 5]
```diff
@@ -335,12 +335,10 @@ class ConvolutionV1Op : public Operator {
                                      oshape[2] * oshape[3]);
     // param_.workspace is in elements of sizeof(DType)
     // if param_.workspace is set to zero the nstep_ equals ishape[0] (batch)
-    nstep_ = std::max(
-        std::min(
-            static_cast<index_t>(
-                param_.workspace / (shape_colunit_.Size() + shape_dstunit_.Size())),
-            ishape[0]),
-        1U);
+    nstep_ = std::max<index_t>(
```
Comment (lebeg) [inline src/operator/convolution_v1-inl.h]: 👍

File: src/operator/nn/upsampling-inl.h [hunk 6]
```diff
@@ -48,8 +48,8 @@ enum UpSamplingMultiInputMode {kConcat, kSum};
 }  // namespace up_enum
 
 struct UpSamplingParam : public dmlc::Parameter<UpSamplingParam> {
-  index_t scale;
-  index_t num_filter;
+  int scale;
+  int num_filter;
```
Comment (lebeg) [inline src/operator/nn/upsampling-inl.h]: `size_t`?

File: src/operator/tensor/ordering_op-inl.h [hunk 7]
```diff
@@ -378,16 +378,19 @@ void TopKImpl(const RunContext &ctx,
     << "The total element_num is " << element_num << ", but the selected IDType can only represent "
     << mxnet::common::MaxIntegerValue<IDType>() << " elements";
   Tensor<xpu, 3, DType> dat = src.FlatTo3D<xpu, DType>(axis, axis, s);
-  size_t temp_size = 0;
+  index_t temp_size = 0;
```
Comment (lebeg) [inline src/operator/tensor/ordering_op-inl.h]: Maybe `size_t` is more appropriate here and the arguments have to be casted instead

File: tests/nightly/test_large_array.py (hunk 2 above)
Comment (lebeg) [reply tests/nightly/test_large_array.py]: I would suggest:
```python
import unittest
class TestLargeArray(unittest.TestCase):
    ...

if __name__ == '__main__':
    unittest.main()
```
And to add a empty line to the end of the file.

File: src/operator/nn/upsampling-inl.h (hunk 6 above)
Comment (apeforest) [reply src/operator/nn/upsampling-inl.h]: I think size_t is usually not preferred in data-structure members, because it has a cross-platform issue

Comment (lebeg) [reply src/operator/nn/upsampling-inl.h]: @apeforest hm, yes, that might be an issue for interoperability. But for implementation use cases it wouldn't matter. `int` or maybe `unsigned int` (or even `uint23_t`) is fine as well.

Comment (apeforest) [reply src/operator/nn/upsampling-inl.h]: Yes, you are right. According to [Google C++ Style Guide](https://google.github.io/styleguide/cppguide.html#Integer_Types) `int` is actually preferred than `unsigned int` in this case.

File: tests/nightly/test_large_array.py (hunk 2 above)
Comment (lebeg) [reply tests/nightly/test_large_array.py]: Well, my suggestion is more about the python code style, where to put it is another topic. Event if it would be a nightly test, you still want to have it in a good shape.

**Comment by apeforest** [comment]
@lebeg Changed unit test style following your suggestion. Please review again. Thanks

**Comment by apeforest** [comment]
@sandeep-krishnamurthy @nswamy Please change the label to [pr-awaiting-review] Thanks!

**Comment by vrakesh** [comment]
There seems to be a regression on resnet-18 model inference time (When running on GPU) post this PR, this was caught in MMS. nightly runs, the changes in this PR seem to be causing this issue.

**Setup**

We use MMS docker images to run load tests, we can start a local container using the following command.

```bash
nvidia-docker run --name mms_benchmark_gpu -p 8080:8080 -p 8081:8081  -itd awsdeeplearningteam/mxnet-model-server:nightly-mxnet-gpu
```

for building MXNet opencv 3.2 and CUDA 9.2 were used. 

Load testing was done using locust, to install locust
```bash
pip install locust
```

# Download Test image
```bash
curl -O https://s3.amazonaws.com/model-server/inputs/kitten.jpg
```
The locust script for load testing
```python
# test_resnet_!8.py
from locust import HttpLocust, TaskSet, task, TaskSequence, seq_task
import urllib
import os
data = None
with open(os.path.join(os.getcwd(),'kitten.jpg'), 'rb') as data:
    data = data.read()

class PredictionTasks(TaskSet):
    @task
    def inference(self):
        self.client.post("/predictions/resnet-18", data=data,headers={'Content-Type': 'image/jpeg'})

class Prediction(HttpLocust):
    task_set = PredictionTasks
    min_wait = 100
    max_wait = 100
```

**Running Load test**

Registering and loading model

```bash
# Register and load resnet-18 model archive
 curl -X POST 127.0.0.1:8081/models?url=https://s3.amazonaws.com/model-server/model_archive_1.0/resnet-18.mar
```

Start a single worker and run latency test

```bash
Start worker and latency test
$ curl -X PUT 'http://127.0.0.1:8081/models/resnet-18?min_worker=1&synchronous=true'
$ locust -f test_resnet_18.py Prediction --host=http://127.0.0.1:8080 --no-web  -c 1 -r 1 -t 20s --only-summary
```

**To change mxnet version/build in docker image,** 

**NOTE** By default recent pip version is pulled.

```bash
# Go into docker image
nvidia-docker exec -u root -it mms_benchmark_gpu bash
$  pip uninstall mxnet-cu92mkl
$ pip install <new-build>.whl
ctrl + p + q to quit docker image

# Destroy existing worker, and create new worker, this loads in newly installed mxnet
$ curl -X PUT 'http://127.0.0.1:8081/models/resnet-18?min_worker=0&synchronous=true'
$ curl -X PUT 'http://127.0.0.1:8081/models/resnet-18?min_worker=1&synchronous=true'
```

**Results**

on mxnet-cu92==1.3.0post0

```bash
# locust result
[2018-12-20 23:34:57,309] ip-172-31-74-13/INFO/locust.main: Run time limit set to 20 seconds
[2018-12-20 23:34:57,310] ip-172-31-74-13/INFO/locust.main: Starting Locust 0.9.0
[2018-12-20 23:34:57,310] ip-172-31-74-13/INFO/locust.runners: Hatching and swarming 1 clients at the rate 1 clients/s...
[2018-12-20 23:34:58,311] ip-172-31-74-13/INFO/locust.runners: All locusts hatched: Prediction: 1
[2018-12-20 23:35:17,310] ip-172-31-74-13/INFO/locust.main: Time limit reached. Stopping Locust.
[2018-12-20 23:35:17,311] ip-172-31-74-13/INFO/locust.main: Shutting down (exit code 0), bye.
[2018-12-20 23:35:17,311] ip-172-31-74-13/INFO/locust.main: Cleaning up runner...
[2018-12-20 23:35:17,311] ip-172-31-74-13/INFO/locust.main: Running teardowns...
 Name                                                          # reqs      # fails     Avg     Min     Max  |  Median   req/s
--------------------------------------------------------------------------------------------------------------------------------------------
 POST /predictions/resnet-18                                      152     0(0.00%)      31      30      39  |      31    7.60
--------------------------------------------------------------------------------------------------------------------------------------------
 Total                                                            152     0(0.00%)                                       7.60

Percentage of the requests completed within given times
 Name                                                           # reqs    50%    66%    75%    80%    90%    95%    98%    99%   100%
--------------------------------------------------------------------------------------------------------------------------------------------
 POST /predictions/resnet-18                                       152     31     31     31     31     32     33     33     34     280
--------------------------------------------------------------------------------------------------------------------------------------------
 Total                                                             152     31     31     31     31     32     33     33     34     280
```
On mxnet-cu92 with commit https://github.com/apache/incubator-mxnet/commit/f9f74169bb05f85d85dec5991aa5fc9050dec9f6 

```bash
 Name                                                          # reqs      # fails     Avg     Min     Max  |  Median   req/s
--------------------------------------------------------------------------------------------------------------------------------------------
 POST /predictions/resnet-18                                      141     0(0.00%)      41      37     337  |      38    7.20
--------------------------------------------------------------------------------------------------------------------------------------------
 Total                                                            141     0(0.00%)                                       7.20

Percentage of the requests completed within given times
 Name                                                           # reqs    50%    66%    75%    80%    90%    95%    98%    99%   100%
--------------------------------------------------------------------------------------------------------------------------------------------
 POST /predictions/resnet-18                                       141     38     39     39     40     40     42     49     49    340
--------------------------------------------------------------------------------------------------------------------------------------------
 Total                                                             141     38     39     39     40     40     42     49     49    340
```

This regression thus carries over to 1.3.1

There is a 30% increase in latency/inference time for resnet-18 based on the above results.

**Comment by TaoLv** [comment]
@apeforest @marcoabreu  Can you point me to the tested history of the new test_large_ndarray.py script? Cannot find it in Jenkins.

**Comment by apeforest** [comment]
@TaoLv It's in tests/nightly/test_large_array.py

**Comment by TaoLv** [comment]
@apeforest Yes, I know the script is there. I'm just wondering if it's actually executed nightly and where can I find the execution history of it.

**Comment by apeforest** [comment]
@marcoabreu @lebeg Could you please confirm if nightly tests under tests/nightly are being run? and where can we find the execution history? Thanks!

**Comment by apeforest** [comment]
@vrakesh Can you open an issue to track this performance degradation with reference to this PR. Adding comment in a merged PR can hardly get noticed. Thanks!
